python/taichi/lang/source_builder.py

```diff
@@ -0,0 +1,69 @@
+import atexit
+import ctypes
+import os
+import tempfile
+
+from taichi.lang.exception import TaichiSyntaxError
+from taichi.lang.ops import asm, bitcode_func_call, external_func_call
+from taichi.lang.util import get_clangpp, has_clangpp
+
+
+class SourceBuilder():
+    def __init__(self, source, mode='bc'):
+        self.mode = mode
+        if self.mode == 'bc':
+            self.td = tempfile.mkdtemp()
+            self.source_file = os.path.join(self.td, 'source.cpp')
+            self.compiled_file = os.path.join(self.td, 'source.bc')
+            with open(self.source_file, 'w') as f:
+                f.write(source)
+            assert has_clangpp()
+            os.system(get_clangpp() + ' -flto -c ' + self.source_file +
```

Comment:
It's better if the user can pass in a lambda that does the compilation (and we provide a default impl if they don't)